moped-database/hasura-cluster

```diff
@@ -39,33 +70,187 @@ function __get_version() {
+# Checks if there is a link, it so it exits
+#
+function __exit_on_link() {
+  LOCAL_PORT=$(__config_get_key "bastion.local_port");
+  if [[ $(__is_local_port_open $LOCAL_PORT) = "TRUE" ]]; then
+    echo "A link cannot be present for this operation, please run './hasura-cluster unlink'";
+    exit 1;
   fi;
 }
 
+#
+# Checks if there is VPN access
+#
+function __check_vpn_access() {
+  curl -skIL --max-time 3 "https://$(__config_get_key "bastion.ssh_bastion")" && echo "TRUE" || echo "FALSE";
+}
+
+#
+# Restores a snapshot into local
+#
+function __restore_snapshot() {
+
+  # Be sure there is no link
+  __exit_on_link;
+
+  DB_SNAPSHOT=$1;
+  DB_SS_BASENAME="${DB_SNAPSHOT:10:512}";
+  echo "Basename: ${DB_SS_BASENAME}";
+
+  docker run -it --rm --network host -v "$(pwd)"/snapshots:/snapshots \
+     --workdir="/snapshots" \
+     --env PGHOST="docker.for.mac.host.internal" \
+     --env PGPORT="5432" \
+     --env PGUSER="moped" \
+     --env PGPASSWORD="moped" \
+     --env PGDATABASE="moped" \
+      $HASURA_CLUSTER_POSTGRES_IMAGE psql -f $DB_SS_BASENAME;
+}
+
+#
+# Clones from remote
+#
+function __create_snapshot() {
+  DB_PROFILE=$1;
+  TIMESTAMP=$(__get_timestamp);
+
+  echo "Creating snapshot from postgres, profile: ${DB_PROFILE}"
+
+  # Exit if there is no link!
+  __check_link_or_exit;
+
+  # Delegate
+  echo "--------------------------------------------------------------------";
+  echo " ⏳ Please seat back and relax, this is going to take a while...";
+  echo "    ** It takes time for pg_dump to download the search_path. **";
+  echo "--------------------------------------------------------------------";
+  sleep 5;
+
+  docker run -it --rm --network host --name moped_download_db_snapshot \
+    -v "$(pwd)"/snapshots:/snapshots \
+    --env PGHOST=$(__config_get_key "bastion.container_parent") \
+    --env PGPORT=$(__config_get_key "bastion.local_port") \
+    --env PGUSER=$(__config_get_key "postgres.${DB_PROFILE}.db_user") \
+    --env PGPASSWORD=$(__config_get_key "postgres.${DB_PROFILE}.db_pass") \
+    --env PGDATABASE=$(__config_get_key "postgres.${DB_PROFILE}.db_name") \
+    $HASURA_CLUSTER_POSTGRES_IMAGE pg_dump --verbose --no-owner --no-privileges --clean \
+    --file="/snapshots/${DB_PROFILE}/postgres_${DB_PROFILE}_${TIMESTAMP}.psql";
+}
+
+#
+# Download a snapshot from bucket
+#
+function __create_snapshot_from_bucket() {
+  LOCAL_ENV=$1;
+  S3_BACKUP_DATE=$2;
+
+  # Check we have AWS
+  __aws_installed_or_exit;
+
+  BUCKET_NAME=$(__config_get_key "backups.${LOCAL_ENV}.bucket");
+  BUCKET_PREFIX=$(__config_get_key "backups.${LOCAL_ENV}.prefix" | sed -r "s/_staging\//_staging/g"| sed -r "s/_production\//_production/g");
+  LOCAL_FILE_PATH="./snapshots/${LOCAL_ENV}/atd_moped_${LOCAL_ENV}-full.sql.gz";
+  LOCAL_FILE_PATH_UNZIPPED="./snapshots/${LOCAL_ENV}/bucket_${LOCAL_ENV}-${S3_BACKUP_DATE}.psql";
+  aws s3 cp "s3://${BUCKET_NAME}/${BUCKET_PREFIX}/${S3_BACKUP_DATE}/atd_moped_${LOCAL_ENV}-full.sql.gz" "${LOCAL_FILE_PATH}";
```

Comment:
Currently, the back up job that is making snapshots for the production database is using the naming scheme `atd_moped-full.sql.gz` where this script expects `atd_moped_production.sql.gz`. I'm not sure if the best solution is to change this script to swap in `full` for `production` when someone is pulling from a production backup, or to change the backup job to name those files in the way the script currently expects them.

Reply:
I think it's a valid observation @frankhereford; however, *.gz files are treated as temporary files and are removed after the execution is finished (see line 224: `rm "${LOCAL_FILE_PATH}";` where LOCAL_FILE_PATH is the zipped file).

In short, the name of the gz file is trivial because it gets removed. Would you agree?

Reply:
Ah, that is a good catch. Thanks @frankhereford